syncer/syncer.go

```diff
@@ -0,0 +1,285 @@
+	SyncerProto      = "/syncer/0.2"
+)
+
+var (
+	errTimeout = errors.New("timeout awaiting block from peer")
+)
+
+// XXX: Don't use this syncer for the consensus that may cause fork.
+// This syncer doesn't assume fork. Consensus may be broken.
+// TODO: Add extensibility for fork before merge
+type syncer struct {
+	logger          hclog.Logger
+	blockchain      Blockchain
+	syncProgression Progression
+
+	peerMap         *PeerMap
+	syncPeerService SyncPeerService
+	syncPeerClient  SyncPeerClient
+
+	// Timeout for syncing a block
+	blockTimeout time.Duration
+
+	// Channel to notify WatchSync that a new status arrived
+	newStatusCh chan struct{}
+}
+
+func NewSyncer(
+	logger hclog.Logger,
+	network Network,
+	blockchain Blockchain,
+	blockTimeout time.Duration,
+) Syncer {
+	return &syncer{
+		logger:          logger.Named(syncerLoggerName),
+		blockchain:      blockchain,
+		syncProgression: progress.NewProgressionWrapper(progress.ChainSyncBulk),
+		syncPeerService: NewSyncPeerService(network, blockchain),
+		syncPeerClient:  NewSyncPeerClient(logger, network, blockchain),
+		blockTimeout:    blockTimeout,
+		newStatusCh:     make(chan struct{}),
+		peerMap:         new(PeerMap),
+	}
+}
+
+// Start starts goroutine processes
+func (s *syncer) Start() error {
+	if err := s.syncPeerClient.Start(); err != nil {
+		return err
+	}
+
+	s.syncPeerService.Start()
+
+	go s.initializePeerMap()
+	go s.startPeerConnectionEventProcess()
+
+	return nil
+}
+
+// Close terminates goroutine processes
+func (s *syncer) Close() {
+	close(s.newStatusCh)
+}
+
+// initializePeerMap fetches peer statuses and initializes map
+func (s *syncer) initializePeerMap() {
+	peerStatuses := s.syncPeerClient.GetConnectedPeerStatuses()
+	s.peerMap.PutPeers(peerStatuses)
+
+	for peerStatus := range s.syncPeerClient.GetPeerStatusUpdateCh() {
+		s.peerMap.Put(peerStatus)
+
+		select {
+		case s.newStatusCh <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// startPeerConnectionEventProcess processes peer connection change events
+func (s *syncer) startPeerConnectionEventProcess() {
+	for e := range s.syncPeerClient.GetPeerConnectionUpdateEventCh() {
+		peerID := e.PeerID
+
+		switch e.Type {
+		case event.PeerConnected:
+			go func() {
+				status, err := s.syncPeerClient.GetPeerStatus(peerID)
+				if err != nil {
+					s.logger.Warn("failed to get peer status, skip", "id", peerID, "err", err)
+
+					return
+				}
+
+				s.peerMap.Put(status)
+
+				select {
+				case s.newStatusCh <- struct{}{}:
+				default:
+				}
+			}()
+		case event.PeerDisconnected:
+			s.peerMap.Remove(peerID)
+		}
+	}
+}
+
+// GetSyncProgression returns progression
+func (s *syncer) GetSyncProgression() *progress.Progression {
+	return s.syncProgression.GetProgression()
+}
+
+// HasSyncPeer returns whether syncer has the peer to syncs blocks
+// return false if syncer has no peer whose latest block height doesn't exceed local height
+func (s *syncer) HasSyncPeer() bool {
+	if s.peerMap == nil {
```

Comment:
This check seems redundant - you initialize the map upon constructing the syncer. Also, the field itself does not have to be a pointer (`*PeerMap`).

Reply:
Good catch, fixed at 693b4791eb446e76c4732b468375c4059a20b816